src/textual/command_palette.py

```diff
@@ -0,0 +1,754 @@
+
+        This method is used to allow clicking on the 'background' as a
+        method of dismissing the palette.
+        """
+        if self.get_widget_at(event.screen_x, event.screen_y)[0] is self:
+            self.workers.cancel_all()
+            self.dismiss()
+
+    def _on_mount(self, _: Mount) -> None:
+        """Capture the calling screen."""
+        self._calling_screen = self.app.screen_stack[-2]
+
+    def _stop_busy_countdown(self) -> None:
+        """Stop any busy countdown that's in effect."""
+        if self._busy_timer is not None:
+            self._busy_timer.stop()
+            self._busy_timer = None
+
+    def _start_busy_countdown(self) -> None:
+        """Start a countdown to showing that we're busy searching."""
+        self._stop_busy_countdown()
+
+        def _become_busy() -> None:
+            if self._list_visible:
+                self._show_busy = True
+
+        self._busy_timer = self._busy_timer = self.set_timer(0.5, _become_busy)
+
+    def _watch__list_visible(self) -> None:
+        """React to the list visible flag being toggled."""
+        self.query_one(CommandList).set_class(self._list_visible, "--visible")
+        self.query_one("#--input", Horizontal).set_class(
+            self._list_visible, "--list-visible"
+        )
+        if not self._list_visible:
+            self._show_busy = False
+
+    async def _watch__show_busy(self) -> None:
+        """React to the show busy flag being toggled.
+
+        This watcher adds or removes a busy indication depending on the
+        flag's state.
+        """
+        self.query_one(LoadingIndicator).set_class(self._show_busy, "--visible")
+        self.query_one(CommandList).set_class(self._show_busy, "--populating")
+
+    @staticmethod
+    async def _consume(
+        source: CommandMatches, commands: Queue[CommandSourceHit]
+    ) -> None:
+        """Consume a source of matching commands, feeding the given command queue.
+
+        Args:
+            source: The source to consume.
+            commands: The command queue to feed.
+        """
+        async for hit in source:
+            await commands.put(hit)
+
+    async def _search_for(self, search_value: str) -> CommandMatches:
+        """Search for a given search value amongst all of the command sources.
+
+        Args:
+            search_value: The value to search for.
+
+        Yields:
+            The hits made amongst the registered command sources.
+        """
+
+        # Get the style for highlighted parts of a hit match.
+        match_style = self._sans_background(
+            self.get_component_rich_style("command-palette--highlight")
+        )
+
+        # Set up a queue to stream in the command hits from all the sources.
+        commands: Queue[CommandSourceHit] = Queue()
+
+        # Fire up an instance of each command source, inside a task, and
+        # have them go start looking for matches.
+        searches = [
+            create_task(
```

Comment:
It doesn't look like these tasks are ever cancelled when a new search comes in. This may not produce an error but it is likely that they will be doing work that is never seen, and I suspect it will generate asyncio warnings.  Might be worth try this with [asyncio debug](https://docs.python.org/3/library/asyncio-dev.html) enabled.